include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -410,10 +410,21 @@ class transform_iterator
     typedef typename ::std::iterator_traits<_Iter>::pointer pointer;
     typedef typename ::std::iterator_traits<_Iter>::iterator_category iterator_category;
 
-    transform_iterator(_Iter __it = _Iter(), _UnaryFunc __unary_func = _UnaryFunc())
-        : __my_it_(__it), __my_unary_func_(__unary_func)
+    //only enable this constructor if both template types are default constructible
```

Comment:
What if we will write something like
```C++
 transform_iterator() = default;

 transform_iterator(_Iter __it)
        : __my_it_(__it)
    {
    }

 transform_iterator(_Iter __it, _UnaryFunc __unary_func)
        : __my_it_(__it), __my_unary_func_(__unary_func)
    {
    }
```

?

Reply:
Sadly it is not enough.  With these formulations `std::is_default_constructible_v<transform_iterator<it,non_default_constructible_functor>>` returns `true`.  It seems that only the immediate context it evaluated during this check, which does not include the "body" of the constructor in question.  We need to explicitly disable the ctor, so that it is not included in the instantiation.

Reply:
OK